burr/core/application.py

```diff
@@ -1127,38 +1181,152 @@ def with_hooks(self, *adapters: LifecycleAdapter) -> "ApplicationBuilder":
+        self.start = default_entry_point
+        return self
+
+    def with_state_persister(
+        self, persister: BasicStatePersistence, on_every: str = "step"
+    ) -> "ApplicationBuilder":
+        """Adds a state persister to the application. This is a way to persist state out to a database, file, etc...
+        at the specified interval.
+
+        :param persister: The persister to add
+        :param on_every: The interval to persist state. Currently only "step" is supported.
+        :return: The application builder for future chaining.
+        """
+        if on_every != "step":
+            raise ValueError(f"on_every {on_every} not supported")
+        # stick persister into a post_step_hook if it doesn't implement it already.
+        if hasattr(persister, "post_step_hook"):
+            if persister not in self.lifecycle_adapters:
+                self.lifecycle_adapters.append(persister)
         else:
-            raise ValueError(f"Tracker {tracker} not supported")
+            # TODO: pull this out more formally
+            class persisterWrapper(PostRunStepHook):
+                def post_run_step(
+                    self,
+                    app_id: str,
+                    partition_key: str,
+                    sequence_id: int,
+                    state: "State",
+                    action: "Action",
+                    result: Optional[Dict[str, Any]],
+                    exception: Exception,
+                    **future_kwargs: Any,
+                ):
+                    if exception is None:
+                        persister.save(
+                            partition_key, app_id, sequence_id, action.name, state, "completed"
+                        )
+                    else:
+                        persister.save(
+                            partition_key, app_id, sequence_id, action.name, state, "failed"
+                        )
+
+            self.lifecycle_adapters.append(persisterWrapper())
         return self
 
+    def _load_from_persister(self):
+        """Loads from the set persister and into this current object.
+
+        Mutates:
+         - self.state
+         - self.sequence_id
+         - maybe self.start
+
+        """
+        # load state from persister
+        load_result = self.persister.load(self.partition_key, self.app_id, self.sequence_id)
+        if load_result is None:
+            # there was nothing to load -- use default state
+            self.state = self.state.update(**self.default_state)
+            self.sequence_id = 0
+        else:
+            # there was something
+            last_position = load_result["position"]
+            self.state = load_result["state"]
+            self.sequence_id = load_result["sequence_id"]
+            status = load_result["status"]
+            if self.resume_at_next_action:
+                # if we're supposed to resume where we saved from
+                if status == "completed":
```

Comment:
The more I'm looking through this the more I think that state persistence should *not* know about whether it errored. How will it change which state to load? Do we even use it? It just sticks a ton of complexity here.

Reply:
Also I don't think we should change `self.start` -- then we forget where we *should* start from, right? That's a fixed property, not something that's dynamic. We use `previous_step` + might need some ability to reset.